examples/yacat/yacat.py

```diff
@@ -20,146 +24,184 @@
     TEXT_COLOR_YELLOW,
 )
 
+HASHCAT_ATTACK_MODE = 3  # stands for mask attack, hashcat -a option
+KEYSPACE_OUTPUT_PATH = Path("/golem/output/keyspace")
 
-def write_hash(hash):
-    with open("in.hash", "w") as f:
-        f.write(hash)
+# Ideally, this value should depend on the chunk size
+ATTACK_TIMEOUT: timedelta = timedelta(minutes=30)
+KEYSPACE_TIMEOUT: timedelta = timedelta(minutes=10)
 
+arg_parser = build_parser("Run a hashcat attack (mask mode) on Golem network")
+arg_parser.add_argument("--hash", type=str, help="Target hash to be cracked", required=True)
+arg_parser.add_argument(
+    "--mask", type=str, help="Hashcat mask to be used for the attack", required=True
+)
+arg_parser.add_argument(
+    "--chunk-size",  # affects skip and limit hashcat parameters
+    type=int,
+    help="Limit for the number of words to be checked as part of a single activity",
+    default=4096,
+)
+arg_parser.add_argument(
+    "--hash-type",
+    type=int,
+    help="Type of hashing algorithm to use (hashcat -m option)",
+    default=400,
+)
+arg_parser.add_argument(
+    "--max-workers",
+    type=int,
+    help="The maximum number of nodes we want to perform the attack on (default is dynamic)",
+    default=None,
+)
+
+# Container object for parsed arguments
+args = argparse.Namespace()
+
+
+async def compute_keyspace(context: WorkContext, tasks: AsyncIterable[Task]):
+    """Worker script which computes the size of the keyspace for the mask attack.
+
+    This function is used as the `worker` parameter to `Golem#execute_tasks`.
+    It represents a sequence of commands to be executed on a remote provider node.
+    """
+    async for task in tasks:
+        cmd = f"hashcat --keyspace " f"-a {HASHCAT_ATTACK_MODE} -m {args.hash_type} {args.mask}"
+        context.run("/bin/bash", "-c", cmd)
+
+        future_result = yield context.commit(timeout=KEYSPACE_TIMEOUT)
+        # each item is the result of a single command on the provider (including setup commands)
+        result: List[CommandExecuted] = await future_result
+        # we take the last item since it's the last command that was executed on the provider
+        cmd_result: CommandExecuted = result[-1]
+
+        if cmd_result.success:
```

Comment:
No need to check if `cmd_result.success == True` here, if it weren't then the line `future_result = yield context.commit(timeout=KEYSPACE_TIMEOUT)` would raise `yapapi.executor.CommandExecutionError`.

Reply:
I'd rewrite the body of `compute_keyspace()` as follows:
```python
    async for task in tasks:
        cmd = f"xhashcat --keyspace " f"-a {HASHCAT_ATTACK_MODE} -m {args.hash_type} {args.mask}"
        context.run("/bin/bash", "-c", cmd)

        try:
            future_result = yield context.commit(timeout=KEYSPACE_TIMEOUT)
            # each item is the result of a single command on the provider (including setup commands)
            result: List[CommandExecuted] = await future_result
            # we take the last item since it's the last command that was executed on the provider
            cmd_result: CommandExecuted = result[-1]
            keyspace = int(cmd_result.stdout)
            task.accept_result(result=keyspace)
        except CommandExecutionError as e:
            print(f"{TEXT_COLOR_RED}Failed to compute attack keyspace: {e}{TEXT_COLOR_DEFAULT}")
            raise RuntimeError(f"Failed to compute attack keyspace: {e}")
```

Reply:
That would print for example:
```
[2021-05-27 13:07:04,963 WARNING yapapi.summary] Worker for provider 'provider-2' failed; reason: Failed to compute attack keyspace: Command '{'run': {'entry_point': '/bin/bash', 'args': ('-c', 'xhashcat --keyspace -a 3 -m 400 ?a?a?a'), 'capture': {'stdout': {'stream': {}}, 'stderr': {'stream': {}}}}}' failed on provider; message: 'ExeScript command exited with code 127'; stderr: 'bash: xhashcat: command not found
```

Reply:
Good point, I'll switch to a try-except in `compute_keyspace`.